core/src/main/java/com/scalar/db/util/ActiveExpiringMap.java

```diff
@@ -65,28 +67,39 @@ private void startValueExpirationThread() {
   }
 
   public Optional<V> get(K key) {
-    if (!map.containsKey(key)) {
+    ValueHolder value = map.get(key);
+    if (value == null) {
       return Optional.empty();
     }
-    ValueHolder<V> value = map.get(key);
     value.updateExpirationTime();
     return Optional.of(value.get());
   }
 
-  public V putIfAbsent(K key, V value) {
-    ValueHolder<V> prev = map.putIfAbsent(key, new ValueHolder<>(value, valueLifetimeMillis));
-    if (prev == null) {
-      return null;
-    }
-    return prev.get();
+  public Optional<V> putIfAbsent(K key, V value) {
+    ValueHolder prev = map.putIfAbsent(key, new ValueHolder(value));
+    return prev == null ? Optional.empty() : Optional.of(prev.get());
   }
 
-  public V put(K key, V value) {
-    ValueHolder<V> prev = map.put(key, new ValueHolder<>(value, valueLifetimeMillis));
-    if (prev == null) {
-      return null;
-    }
-    return prev.get();
+  public Optional<V> put(K key, V value) {
+    ValueHolder prev = map.put(key, new ValueHolder(value));
+    return prev == null ? Optional.empty() : Optional.of(prev.get());
+  }
+
+  public Optional<V> computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
+    ValueHolder prev = map.computeIfAbsent(key, mappingFunction.andThen(ValueHolder::new));
+    return prev == null ? Optional.empty() : Optional.of(prev.get());
+  }
+
+  public Optional<V> computeIfPresent(
+      K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
+    ValueHolder prev =
+        map.computeIfPresent(
+            key,
+            (k, oldValue) -> {
+              V newValue = remappingFunction.apply(k, oldValue.get());
+              return new ValueHolder(newValue);
+            });
+    return prev == null ? Optional.empty() : Optional.of(prev.get());
```

Comment:
Added `computeIfAbsent()` and `computeIfPresent()` methods to `ActiveExpiringMap`. This is a main change.